bgpd: fix consider MAX PACKET SIZE when receiving peer has option

pguibert6WIND opened the pull request:
When peering between two eBGP peers initiated by A, if BGP from B sends a lot of BGP updates, A sends back the updates in a big update packet, and that packet is rejected.

> 2025/08/19 18:37:23 BGP: [G00A3-MV858] 192.168.0.1 bad message length - 32408 for UPDATE
> 2025/08/19 18:37:23 BGP: [HZN6M-XRM1G] %NOTIFICATION: sent to neighbor 192.168.0.1 1/2 (Message Header Error/Bad Message Length) 2 bytes 7

Actually, B device considers A has a standard packet size of 4096, whereas A already had sent in its open message the info that packet size of 65535 are authorised.

Fix this by only relying on the incoming open message, as by default, BGP instance sends BGP open messages with the instruction.

Fixes: 8d976b0e2b78 ("bgpd: Set extended msg size only if we advertised and received capability")

ton31337 commented:
How the original issue will be solved with this revert?

pguibert6WIND committed:
Revert "bgpd: Set extended msg size only if we advertised and received capability"

When peering between two eBGP peers initiated by A, if BGP from B sends
a lot of BGP updates, A sends back the updates in a big update packet,
and that packet is rejected.

> 2025/08/19 18:37:23 BGP: [G00A3-MV858] 192.168.0.1 bad message length - 32408 for UPDATE
> 2025/08/19 18:37:23 BGP: [HZN6M-XRM1G] %NOTIFICATION: sent to neighbor 192.168.0.1 1/2 (Message Header Error/Bad Message Length) 2 bytes 7

Actually, B device considers A has a standard packet size of 4096,
whereas A already had sent in its open message the info that packet size
of 65535 are authorised. Revert the original commit.

This reverts commit 8d976b0e2b78c3755d8d7868c535db8c8a163008.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
bgpd: set max_packet_size to MAX if dont-capability-negotiate is not set

The original commit was wrongly looking at ADV value, which is not set
when BGP device is responder. The only case where the STANDARD value
is used, is when dont-capability-negotiate is used.

Fix this by adding the condition.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND commented:
You are right. I did not check well about the original pull request.
Done the revert and submitted a fix

pguibert6WIND commented on bgpd/bgp_open.c:
```diff
@@ -1475,12 +1475,13 @@ int bgp_open_option_parse(struct peer *peer, uint16_t length,
+	 * if peer is capable and dont-capability-negotiate is not set
+	 */
+	peer->max_packet_size = CHECK_FLAG(peer->cap, PEER_CAP_EXTENDED_MESSAGE_RCV) &&
+						!CHECK_FLAG(peer->flags, PEER_FLAG_DONT_CAPABILITY)
```
My understanding is the following one:
The MAX message support size is an option sent in BGP open only if capability-negotiate is true.
The MAX message support is always true when set.

So coming back to A and B.
As A first sends a BGP open to B, then B receives the message. B did not yet forge the BGP reply open message, so the ADV flag has not been set, and I can only rely on the peer configuration (from B config) which I expect to be in the peer structure: I expect dont-capability-negotiate is not set, in order to set MAX message to BGP open too.

pguibert6WIND commented:
> it sure seems like this could be a topotest with some sort of show commands on either side showing the max packet size being different for the two sides, right? I too would like to understand the sequence of events. I don't understand what is going wrong here yet either

I can share a private branch I used for a separate problem.
a toptoest is available.
``
python3 -m pytest bgp_bmp/test_bgp_bmp_4.py -vs --pause
```

branch:
https://github.com/FRRouting/frr/compare/master...pguibert6WIND:frr:trial_bmp_queues

riw777 reviewed (approved):
I think, based on the current explanation and notes, this is a good fix ... I'm with @ton31337 in that I don't see how this is different from the original (reverted) fix.

The logic here in the BGP state machine is not the easiest to follow.

pguibert6WIND commented:
> show ip bgp update-group

```
r2# show ip bgp update-group
Update-group 1:
  Created: Tue Sep  2 16:36:41 2025
  MRAI value (seconds): 0

  Update-subgroup 1:
    Created: Tue Sep  2 16:36:41 2025
    Join events: 1
    Prune events: 0
    Merge events: 0
    Split events: 0
    Update group switch events: 0
    Peer refreshes combined: 0
    Merge checks triggered: 0
    Coalesce Time: 1150
    Version: 1
    Packet queue length: 0
    Total packets enqueued: 1
    Packet queue high watermark: 1
    Adj-out list count: 1
    Advertise list: empty
    Flags: 
    Max packet size: 4096
    Peers:
      - 192.168.0.1
r2# 

r1multibmp# show ip bgp update-group
Update-group 1:
  Created: Tue Sep  2 16:36:41 2025
  MRAI value (seconds): 0

  Update-subgroup 1:
    Created: Tue Sep  2 16:36:41 2025
    Join events: 1
    Prune events: 0
    Merge events: 0
    Split events: 0
    Update group switch events: 0
    Peer refreshes combined: 0
    Merge checks triggered: 0
    Coalesce Time: 1150
    Version: 1
    Packet queue length: 0
    Total packets enqueued: 1
    Packet queue high watermark: 1
    Adj-out list count: 1
    Advertise list: empty
    Flags:
    Max packet size: 65535
    Peers:
      - 192.168.0.2
r1multibmp# 
```

ton31337 commented:
IMNSHO, it would be cleaner if we had `neighbor X capability extended-message` flag (instead of relying on `dont-capability`), what do you think?

vjardin commented:
@pguibert6WIND : do you plan to add this command ?

riw777 commented:
@greptile review
